kinfer/rust/kinfer/src/lib.rs

```diff
@@ -0,0 +1,240 @@
+        let input_map: HashMap<_, _> = input_values.into_iter().collect();
+        
+        // Run inference
+        let outputs = self.session.run(&input_map)?;
+
+        // Convert output format to match input format
+        match inputs {
+            InputData::Single(_) => Ok(OutputData::Single(outputs.remove(0))),
+            InputData::List(_) => Ok(OutputData::List(outputs)),
+            InputData::Map(_) => {
+                let named_outputs = self.output_details
+                    .iter()
+                    .zip(outputs)
+                    .map(|(info, tensor)| (info.name.clone(), tensor))
+                    .collect();
+                Ok(OutputData::Map(named_outputs))
+            }
+        }
+    }
+
+    pub fn get_metadata(&self) -> &HashMap<String, JsonValue> {
+        &self.metadata
+    }
+
+    pub fn get_input_details(&self) -> &[TensorInfo] {
+        &self.input_details
+    }
+
+    pub fn get_output_details(&self) -> &[TensorInfo] {
+        &self.output_details
+    }
+}
+
+impl<'a> From<Value<'a>> for InputData<'a> {
+    fn from(tensor: Value<'a>) -> Self {
+        InputData::Single(tensor)
+    }
+}
+
+impl<'a> From<Vec<Value<'a>>> for InputData<'a> {
+    fn from(tensors: Vec<Value<'a>>) -> Self {
+        InputData::List(tensors)
+    }
+}
+
+impl<'a> From<Vec<(&'a str, Value<'a>)>> for InputData<'a> {
+    fn from(named_tensors: Vec<(&'a str, Value<'a>)>) -> Self {
+        InputData::Map(named_tensors)
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use ndarray::{Array, Array2};
+    use ort::{Environment, ExecutionProvider};
+    use std::sync::Arc;
+    use tempfile::NamedTempFile;
+
+    fn create_test_model() -> Result<NamedTempFile, Box<dyn std::error::Error>> {
+        // Create a temporary file that automatically gets cleaned up
+        let model_file = NamedTempFile::new()?;
+        
+        // Use Python to create and export a test model
+        let python_code = format!(
+            r#"
+import torch
+import torch.nn as nn
+import onnx
+
+class SimpleModel(nn.Module):
+    def __init__(self):
+        super().__init__()
+        self.linear = nn.Linear(10, 2)
+    
+    def forward(self, x):
+        return self.linear(x)
+
+# Create model and add metadata
+model = SimpleModel()
+dummy_input = torch.randn(1, 10)
+
+# Add metadata
+metadata = {{"test_key": "test_value"}}
+import json
+meta = onnx.StringStringEntryProto()
+meta.key = "kinfer_metadata"
+meta.value = json.dumps(metadata)
+
+# Export model
+torch.onnx.export(model, dummy_input, '{}')
+model_proto = onnx.load('{}')
+model_proto.metadata_props.append(meta)
+onnx.save(model_proto, '{}')
+"#,
+            model_file.path().display(),
+            model_file.path().display(),
+            model_file.path().display()
+        );
+        
+        let status = std::process::Command::new("python")
+            .arg("-c")
+            .arg(python_code)
+            .status()?;
+            
```

Comment:
this will be annoying to run as a stand-alone rust test

we should just save a sample `test.kmodel` file in the `tests` directory and use that as our reference in the unit tests